### PR TITLE
emacs: Add infrastructure for native-comp

### PR DESCRIPTION
**Pull request by adisbladis**
###### Motivation for this change
This is to facilitate the `emacsGcc` support in https://github.com/nix-community/emacs-overlay.
It's not feasible to add all of this through overrides.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/build-support/emacs/generic.nix [hunk 1]
```diff
@@ -56,6 +56,18 @@ stdenv.mkDerivation ({
   meta = defaultMeta // meta;
 }
 
+// lib.optionalAttrs emacs.nativeComp {
```
Comment (matthewbauer) [inline pkgs/build-support/emacs/generic.nix:59]:
```suggestion
// lib.optionalAttrs (emacs.nativeComp or false) {
```

for emacs from older nixpkgs

Comment (adisbladis) [reply pkgs/build-support/emacs/generic.nix]: Good catch, thanks.

**Comment by antifuchs** [comment]
Thanks for getting this merged! 

I just tried building the native-comp emacs on darwin (which I know works outside nix; most instructions refer to homebrew), and found two errors: One is fixed by 3a79681eb45df9c89d1de613114394f49582f0d6 (which fixes installation of libgccjit), and the other is that https://github.com/NixOS/nixpkgs/pull/93716/files#diff-c88608e19b22610f68cf3cce033c0678R69 depends on `stdenv.glibc`, which doesn't work on Darwin. I'm building a merged set of those two fixes (replacing glibc with libc), and hope to submit a follow-on PR soon: building the world just takes a few hours on this 4-core macbook pro.

File: pkgs/applications/editors/emacs/default.nix [hunk 2]
```diff
@@ -50,29 +56,50 @@ stdenv.mkDerivation rec {
...
+    (lib.optionalString srcRepo ''
+      rm -fr .git
+    '')
+
+    # Make native compilation work both inside and outside of nix build
+    (lib.optionalString nativeComp (let
+      libPath = lib.concatStringsSep ":" [
+        "${lib.getLib libgccjit}/lib/gcc/${targetPlatform.config}/${libgccjit.version}"
+        "${lib.getLib stdenv.cc.cc}/lib"
+        "${lib.getLib stdenv.glibc}/lib"
```
Comment (matthewbauer) [inline pkgs/applications/editors/emacs/default.nix:69]: this should be `stdenv.cc.libc` instead of glibc

**Comment by evanrelf** [comment]
@antifuchs Do you think you'll have a PR up soon to add Darwin support? I'm having trouble coming up with a fix on my own.

**Comment by antifuchs** [comment]
@evanrelf I'm sorry, I didn't manage to make any progress. There's the `.libc` change that was pointed out on the commit; but so far, I'm failing to compile a `libgccjit`. I got the error message "symbols referenced by indirect symbol table entries that can't be stripped" building that library, which apparently means you have to run `strip -x` on macOS... but I couldn't find out where to hook that in.

Maybe we ought to report one or two other issues about Darwin builds there. /:

**Comment by antifuchs** [comment]
@evanrelf after thinking about this more, here's a branch that *seems* to build `libgccjit` and is well on the way towards building the world and then emacs (I'm about halfway through building LLVM right now, all four cores on this machine are sweating). Here's the branch: https://github.com/NixOS/nixpkgs/compare/staging...antifuchs:gccemacs-on-darwin - I hope this helps (and most of all, that it works! We're spending a lot of watts on this)

**Comment by evanrelf** [comment]
@antifuchs Could it be because `stdenv.cc` on Darwin is set to `clang`?

```
nix-repl> pkgs.stdenv.cc.cc.name
"clang-7.1.0"
```

You probably need to force it to use `gcc` explicitly, rather than `stdenv.cc`.

**Comment by antifuchs** [comment]
@evanrelf my understanding is that you have to compile (nextstep-type) Emacs with clang on darwin because gcc hasn't caught up to the language features yet (or you'd have to build it with X11 GUI - [see](https://gist.github.com/mikroskeem/0a5c909c1880408adf732ceba6d3f9ab#gistcomment-3282250)). I think the build dependencies up to libgccjit work, except the libgccjit dynamic library is installed as `libgccjit.so.0.0.1`, not as `libgccjit.dylib`. Emacs's configure script builds the test program with `-lgccjit`, and the resulting binary (according to `otool -L` doesn't have the path to the dynamic library coded into it, and fails with "Image not found".

I guess I have to figure out how to install it as a .dylib... or something else.

**Pull request merged**